### ISSUE TITLE
TypeError: issubclass() arg 1 must be a class

### ISSUE THREAD

**Issue by Casyfill** [samuelcolvin/pydantic#1298]
# Bug

Getting issubcluss error on /docs in Fastapi

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.4
            pydantic compiled: False
                 install path: /Users/philippk/anaconda3/envs/py37/lib/python3.7/site-packages/pydantic
               python version: 3.7.3 | packaged by conda-forge | (default, Dec  6 2019, 08:36:57)  [Clang 9.0.0 (tags/RELEASE_900/final)]
                     platform: Darwin-19.2.0-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
(py37) ZG00021:se-sestimate philippk$ docker run -it -p 80:80 -e ENVKEY -e SNOWFLAKE_PASSWORD='dummy' -e SNOWFLAKE_ACCOUNT='zillow' -e SNOWFLAKE_USER='philippk' -e SNOWFLAKE_WAREHOUSE='LOOKER_WH' -e SNOWFLAKE_DATABASE='FEATURESTORE' -e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY -e AWS_SESSION_TOKEN -e AWS_SECURITY_TOKEN --entrypoint /bin/bash se-sestimate:local-api
root@d50a194c1238:/app# python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.4
            pydantic compiled: True
                 install path: /usr/local/lib/python3.7/site-packages/pydantic
               python version: 3.7.4 (default, Sep 12 2019, 15:40:15)  [GCC 8.3.0]
                     platform: Linux-4.19.76-linuxkit-x86_64-with-debian-10.1
     optional deps. installed: []
```
Here is the code of request_models:

```
from typing import List
from pydantic import BaseModel
from enum import Enum
from datetime import date


class ModelName(str, Enum):
    repeated_sale = "repeated_sale"
    main = "main"
    dummy = "dummy"


class Inference(BaseModel):
    price: int
    model: ModelName
    model_version: float
    date: date

@app.get("/infer", response_model=Inference)
def infer(
    params: dict,
    model: ModelName,
    v: Optional[float] = None,
    date: Optional["date"] = None,
) -> Inference:
    """attempts to pull predictions from snowflake"""
    result = app.model_registry.pull(model=model, v=v, date=date).predict(params)
    return {"model": model, "v": v, **result}
```

this results in:
```
  File "/Users/philippk/anaconda3/envs/py37/lib/python3.7/site-packages/pydantic/schema.py", line 657, in field_singleton_schema
    if issubclass(field_type, Enum):
TypeError: issubclass() arg 1 must be a class
```

**Comment by dmontagu** [comment]
So, I was able to fix this by changing `Optional["date"]` to `Optional[date]`. The problem is that the forward ref inside the `Optional` isn't getting resolved.

I'm not sure whether the problem should be considered to lie in fastapi or pydantic, but it appears to be stemming from an inability for `pydantic.schema.field_schema` to handle a field with type `Optional[ForwardRef("date")]` (presumably it would struggle to handle fields with any `ForwardRef`s).

@samuelcolvin Should FastAPI be taking responsibility for ensuring fields don't have forward refs in their types?

**Comment by samuelcolvin** [comment]
Does `update_forward_refs` fix this?

If so, I think the problem is on the FastAPI side. 

Otherwise, we might be able to have a crack at resolving it in pydantic.

**Comment by half2me** [comment]
I'm having the exact same issue. I have 3 models which reference each other in a circular fashion. I'm using `from __future__ import annotations` to handle forward refs. How can I resolve this issue?

**Comment by half2me** [comment]
Simple example of the issue. Try to access /docs

```python
from typing import Set

from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class A(BaseModel):
    b: Set['B'] = () # if this line is commented out, /docs works fine


class B(BaseModel):
    a: Set[A] = ()


@app.get("/")
def test(a: A):
    return a
```

**Comment by sarendsen** [comment]
I just experience the same bug. update_forward_refs does not solve it.

**Comment by hmalphettes** [comment]
> Simple example of the issue. Try to access /docs on FastAPI

`A.update_forward_refs()` does work for me:

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class A(BaseModel):
    b: 'B'


class B(BaseModel):
    a: A

A.update_forward_refs() # This fixes the issubclass crash
# B.update_forward_refs() # This does not make a difference

@app.get("/")
def hello(a: A):
    return a


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8001)

```

**Comment by PrettyWood** [comment]
Hello
I feel like this issue should be closed as the other issues are not really related.
The `issubclass` issue is now solved since we use `lenient_issubclass`.
The second issue comes from how `__annotations__` works.
When writing
```python
class A:
    date: Optional[date] = None
```
the `__annotations__` is `{'date': <class 'NoneType'>}`, which leads to errors!
Unfortunately it has nothing to do with _pydantic_ and this is a tricky error. 

This doesn't happen if we use `Optional['date'] = None` or `Optional[date]` without default value.
Using `from __future__ import annotations` also works as it does the same trick (use `ForwardRef`...) so I feel like this error should be solved once the new annotations system is supported and the default one.
And for this we already have some open issues. Don't know if we need to open a new one 🤷‍♂️

**Comment by thiras** [comment]
@PrettyWood has this fix released? I'm still getting the error without `.update_forward_refs()`

**Comment by samuelcolvin** [comment]
This will be fixed by #2221 which I'm about to merge and should get released very soon.

@thiras since you're obviously eager for a fix, it would be really helpful if you could try #2221 and let us know if it fixes your problem or not.

**Comment by geropellicer** [comment]
I've had this issue a couple of times using Pydantic 1.8.2 and Fastapi 0.70.0 and my traceback usually looks like this:

```
api        | INFO:     172.23.0.1:62132 - "GET /docs HTTP/1.1" 200 OK
api        | INFO:     172.23.0.1:62132 - "GET /api/v1/openapi.json HTTP/1.1" 500 Internal Server Error
api        | ERROR:    Exception in ASGI application
api        | Traceback (most recent call last):
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 373, in run_asgi
api        |     result = await app(self.scope, self.receive, self.send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
api        |     return await self.app(scope, receive, send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 208, in __call__
api        |     await super().__call__(scope, receive, send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/applications.py", line 112, in __call__
api        |     await self.middleware_stack(scope, receive, send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
api        |     raise exc
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
api        |     await self.app(scope, receive, _send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
api        |     raise exc
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
api        |     await self.app(scope, receive, sender)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/routing.py", line 656, in __call__
api        |     await route.handle(scope, receive, send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/routing.py", line 259, in handle
api        |     await self.app(scope, receive, send)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/starlette/routing.py", line 61, in app
api        |     response = await func(request)
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 161, in openapi
api        |     return JSONResponse(self.openapi())
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 136, in openapi
api        |     self.openapi_schema = get_openapi(
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/fastapi/openapi/utils.py", line 387, in get_openapi
api        |     definitions = get_model_definitions(
api        |   File "/opt/pysetup/.venv/lib/python3.9/site-packages/fastapi/utils.py", line 24, in get_model_definitions
api        |     m_schema, m_definitions, m_nested_models = model_process_schema(
api        |   File "pydantic/schema.py", line 548, in pydantic.schema.model_process_schema
api        |   File "pydantic/schema.py", line 589, in pydantic.schema.model_type_schema
api        |   File "pydantic/schema.py", line 241, in pydantic.schema.field_schema
api        |   File "pydantic/schema.py", line 440, in pydantic.schema.field_type_schema
api        |   File "pydantic/schema.py", line 773, in pydantic.schema.field_singleton_schema
api        |   File "pydantic/schema.py", line 667, in pydantic.schema.field_singleton_sub_fields_schema
api        |   File "pydantic/schema.py", line 495, in pydantic.schema.field_type_schema
api        |   File "pydantic/schema.py", line 839, in pydantic.schema.field_singleton_schema
api        |   File "/usr/local/lib/python3.9/abc.py", line 123, in __subclasscheck__
api        |     return _abc_subclasscheck(cls, subclass)
api        | TypeError: issubclass() arg 1 must be a class
```

as @hmalphettes stated, using `update_forward_refs` works fine for me

**Comment by alispa** [comment]
For me `update_forward_refs` sadly doesn't work:

```
from app1 import ItemOut
if TYPE_CHECKING:
    from app2 import WSAOut

class IQALOut(BaseModel):
    id: int
    quantity: float

    item: ItemOut
    wsa: Optional['WSAOut']

IQALOut.update_forward_refs()
```

This gets me to `NameError: name 'PWSABasicOut' is not defined`.
And if I delete the line with  `update_forward_refs()` the server restarts well, but if I try to access the docs `http://localhost:8000/docs` i get:

```
...
  File "C:\Users\Amel\Desktop\fastapi_training\virtual_environments\envamel\lib\site-packages\pydantic\schema.py", line 495, in field_type_schema
    f_schema, f_definitions, f_nested_models = field_singleton_schema(
  File "C:\Users\Amel\Desktop\fastapi_training\virtual_environments\envamel\lib\site-packages\pydantic\schema.py", line 839, in field_singleton_schema
    if issubclass(field_type, BaseModel):
  File "C:\Users\Amel\AppData\Local\Programs\Python\Python310\lib\abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```

Pydantic 1.8.2 and Fastapi 0.70.0

**Comment by alispa** [comment]
I found a solution following the suggested answers.
The difference was just that I had to put my schema imports at the very bottom, after all classes, and then call for every class with a related field `update_forward_refs()`.


Example:

```
from __future__ import annotations
from pydantic import BaseModel

class A(BaseModel):
    my_x: X

class B(BaseModel):
    my_y: Y

class C(BaseModel):
    my_z: int

# related schemas are imported at bottom
from myapp.schemas.x import X
from myapp.schemas.y import Y

A.update_forward_refs()
B.update_forward_refs()
```

My main mistake, following other answers was to put the import immediately after the class

**Comment by aleksul** [comment]
> I would like to share another weird edge case (with a workaround if it can help someone...)
> ### Versions
> 
> Python 3.9.9 pydantic==1.8.2
> ### Repro
> 
> ```python
> from pydantic import BaseModel
> 
> class Nested(BaseModel):
>     nested_list: list["Nested"]
> 
> Nested.update_forward_refs()
> print(Nested(nested_list=[]))
> ```

This edge case should be fixed with PR #3681.
